qmlti/LTI_Data_Connector_qmp.php

```diff
@@ -826,13 +858,119 @@ public function ReportConfig_update($consumer_key, $resource_link_id, $assessmen
+    'WHERE (consumer_key = :key) AND (context_id = :id)';
+    $query = $this->db->prepare($sql);
+    $query->bindValue('key', $consumer_key, PDO::PARAM_STR);
+    $query->bindValue('id', $context_id, PDO::PARAM_STR);
+    $ok = $query->execute();
+    if ($ok) {
+      $users = $query->fetchAll();
+    } else {
+      return FALSE;
+    }
+    return $users;
+  }
 
 ###
-#    Load the user from the database
+#     loads list of users available on Tool Consumer by context
 ###
+
+  public function TCUser_loadUsers($consumer_key) {
+    $sql = 'SELECT user_id, firstname, lastname, fullname, email, roles, created, updated, lti_result_sourcedid ' .
+    'FROM ' . $this->dbTableNamePrefix . LTI_Data_Connector::TC_USER_TABLE_NAME . ' ' .
+    'WHERE (consumer_key = :key)';
+    $query = $this->db->prepare($sql);
+    $query->bindValue('key', $consumer_key, PDO::PARAM_STR);
+    $ok = $query->execute();
+    if ($ok) {
+      $users = $query->fetchAll();
+    } else {
+      return FALSE;
+    }
+    return $users;
+  }
+
+###
+#     loads list of users available on Tool Consumer by context
+###
+
+  public function User_loadUsers($consumer_key) {
+    $sql = 'SELECT user_id, firstname, lastname, fullname, email, roles, created, updated, lti_result_sourcedid ' .
+    'FROM ' . $this->dbTableNamePrefix . LTI_Data_Connector::USER_TABLE_NAME . ' ' .
+    'WHERE (consumer_key = :key)';
+    $query = $this->db->prepare($sql);
+    $query->bindValue('key', $consumer_key, PDO::PARAM_STR);
+    $ok = $query->execute();
+    if ($ok) {
+      $users = $query->fetchAll();
+    } else {
+      return FALSE;
+    }
+    return $users;
+  }
+
+###
+#    Load the user from the database
+###    
+
   public function User_load($user) {
 
-    return TRUE;
+    $key = $user->getResourceLink()->getKey();
+    $id = $user->getContext();
+    $userId = $user->getId(LTI_Tool_Provider::ID_SCOPE_ID_ONLY);
+    $sql = 'SELECT firstname, lastname, fullname, email, roles, created, updated, lti_result_sourcedid ' .
+           'FROM ' . $this->dbTableNamePrefix . LTI_Data_Connector::USER_TABLE_NAME . ' ' .
+           'WHERE (consumer_key = :key) AND (context_id = :id) AND (user_id = :user_id)';
+    $query = $this->db->prepare($sql);
+    $query->bindValue('key', $key, PDO::PARAM_STR);
+    $query->bindValue('id', $id, PDO::PARAM_STR);
+    $query->bindValue('user_id', $userId, PDO::PARAM_STR);
+    $ok = $query->execute();
+    if ($ok) {
+      $row = $query->fetch(PDO::FETCH_ASSOC);
+      $ok = ($row !== FALSE);
+    }
+
+    if ($ok) {
+      $row = array_change_key_case($row);
+      $user->firstname = $row['firstname'];
+      $user->lastname = $row['lastname'];
+      $user->fullname = $row['fullname'];
+      $user->email = $row['email'];
+      $user->roles = explode(", ", $row['roles']);
+      $user->created = strtotime($row['created']);
+      $user->updated = strtotime($row['updated']);
+      $user->lti_result_sourcedid = $row['lti_result_sourcedid'];
```

Comment:
Upon further research, it looks like the $user variable is adjusted on load - when the user is passed in the function call, that same user object is manipulated. This means that we can have the return statement simply for checking whether or not the load was completed properly - however, the load would have already been done in the function itself.

I don't really like this implementation - it should be two functions, not one. I will still refactor this so it makes more sense, but I can just provide some context here 😄